File: 0234_Palindrome_Linked_List/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
// T = O(n), S = O(1);
class Solution {
  public:
    bool isPalindrome(ListNode *head) {
        int count = 0;
        for (ListNode *p = head; p; count++, p = p->next)
            ;
        ListNode *cur = head, *pre = nullptr;
        for (int i = 0; i < count / 2; i++) {
            ListNode *next = cur->next;
            cur->next = pre;
            pre = cur;
            cur = next;
        }
        if (count % 2) {
            cur = cur->next;
        }
        while (cur) {
            if (pre->val != cur->val) {
                return false;
            }
            cur = cur->next;
            pre = pre->next;
        }
        return true;
    }
};

File: 0234_Palindrome_Linked_List/golang/main.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}
	p := head
	n := 0
	for p != nil {
		n++
		p = p.Next
	}
	var pre *ListNode = nil
	p = head
	for i := 0; i < n/2; i++ {
		next := p.Next
		p.Next = pre
		pre = p
		p = next
	}
	if n%2 == 1 {
		p = p.Next
	}
	for pre != nil {
		if pre.Val != p.Val {
			return false
		}
		pre = pre.Next
		p = p.Next
	}
	return true
}